designs/plugins-phase-2.md

```diff
@@ -0,0 +1,105 @@
+# Plugins Phase 2
+
+## Overview
+
+Phase 2 of the scaffold plugin system adds support for chaining plugins together. The idea for plugins phase 2 is to have support for a plugin architecture that enables plugins to be separate binaries discoverable by the Kubebuilder CLI binary via user specified plugin file paths and that they are language indepedent. Some of the phase 2 high-level requirements summarized from the feature request and discussions are: 
+
+* mdbook-like structure
+    * A Plugin receives a JSON blob plugin "universe" from upstream plugins containing file "nodes", each of which contain file metadata.
+    * The plugin can add or delete nodes (nodes are not mutable), then returns the modified universe so downstream plugins can do the same.
+* Pre-generate a project file for handling complex plugin use-cases.
+* Enhanced "layout" PROJECT file key.
+    * Either an ordered list of plugin {name}/{version}, or ordered map of {name}/{version} to plugin-specific config.
+    * Order matters for certain plugins, ex. a bazel plugin would need to run after all Go files are generated.
+
+## Evaluation of Plugin libraries 
+
+##### Go built-in plugin package vs Hashicorp plugin library vs custom plugin library
+
+Based on the evaluation of plugin libraries such as the built-in go-plugin library and Hashicorp’s plugin library, we have come to the conclusion that it is more suitable to write our own custom typed plugin library.
+
+The built-in plugin library seems to be a non-starter as it is more suitable for in-tree plugins rather than out-of-tree and it doesn’t offer cross language support. Hashicorp’s go plugin system seems more suitable than built-in go-plugin library as it enables cross language/platform support. However, it is more suited for long running plugins as opposed to short lived plugins and it could be overkill with the usage of protobuf.
+ 
+For the stated reasons, the proposal is to write our own plugin system that passes JSON blobs back and forth across stdin/stdout.
+The plugin specification will include type metadata to allow the potential for using other plugin libraries in the future if the need arises. From an implementation standpoint, we could create a new implementation of our plugin interface based on the type specified in the project file. We can do a type switch on the type of plugin the data is being passed to. For ex: if it is native go, then we pass the universe directly & if it is a binary wrapper, then we pass the serialized stream of JSON bytes to it. 
```

Comment:
From the implementation viewpoint, instead of making the cli understand all different kind of plugins, we could code a plugin with the current interface that is in charge of doing this translation. For example, we have a JSON plugin that has the current plugin interface and creates the JSON blobs and pass them to the external plugin.

Now, imagine I'm a C++ plugin developer, I would copy this JSON plugin and develop my C++ plugin that interacts with the JSON-based exposed API.

By doing this, we would only need to support out-of-tree Go plugins, as the integration with other languages would be done by this JSON (and potentially others in a future) plugins. What do you hink about this approach? And would this approach bring the built-in plugin library back to the table?

By the way, I think that these plugins should work under demand more than sending all the data. The C++ plugin would request the `admiral_types.go` file representation and the plugin would sent it in a JSON blob. Obviously the plugin will send some general metadata first (for example, but not limitd to, the plugin config).